src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -563,27 +624,231 @@ internal SafeFileHandle OpenPort(string portName)
-        ~SerialStream()
-        {
-            Dispose(false);
+            _ioLoop = Task.Factory.StartNew(IOLoop, TaskCreationOptions.LongRunning);
         }
 
         protected override void Dispose(bool disposing)
         {
-            // Signal the other side that we're closing.  Should do regardless of whether we've called
-            // Close() or not Dispose()
-            if (_handle != null && !_handle.IsInvalid)
+            if (disposing)
             {
-                   Interop.Sys.Shutdown(_handle, SocketShutdown.Both);
-                   _handle.Close();
+                _ioLoopFinished = true;
+                _ioLoop?.Wait();
+                _ioLoop = null;
+
+                while (_readQueue.TryDequeue(out SerialStreamAsyncResult r))
+                {
+                    r.Complete(new ObjectDisposedException(SR.Port_not_open));
+                }
+
+                while (_writeQueue.TryDequeue(out SerialStreamAsyncResult r))
+                {
+                    r.Complete(new ObjectDisposedException(SR.Port_not_open));
+                }
+
+                // Signal the other side that we're closing.  Should do regardless of whether we've called
+                // Close() or not Dispose()
+                if (_handle != null && !_handle.IsInvalid)
+                {
+                    Interop.Sys.Shutdown(_handle, SocketShutdown.Both);
+                    _handle.Close();
                     _handle = null;
-                    base.Dispose(disposing);
-                    if (PinChanged != null || ErrorReceived != null || DataReceived != null)
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        // should return non-negative integer meaning numbers of bytes read/written (0 for errors)
+        private delegate int RequestProcessor(SerialStreamAsyncResult r);
+
+        private unsafe int ProcessRead(SerialStreamAsyncResult r)
+        {
+            Span<byte> buff = r.Buffer.Span;
+            fixed (byte* bufPtr = buff)
+            {
+                // assumes dequeue-ing happens on a single thread
+                int numBytes = Interop.Sys.Read(_handle, bufPtr, buff.Length);
+
+                Interop.Error lastError = numBytes < 0 ? Interop.Sys.GetLastError() : 0;
+                if (numBytes > 0)
+                {
+                    r.Complete(numBytes);
+                    return numBytes;
+                }
+                else if (numBytes == 0)
+                {
+                    ThreadPool.QueueUserWorkItem(s => {
+                            var thisRef = (SerialStream)s;
+                            thisRef.DataReceived(thisRef, new SerialDataReceivedEventArgs(SerialData.Eof));
+                        }, this);
+                }
+                // ignore EWOULDBLOCK since we handle timeout elsewhere
+                else if (lastError != Interop.Error.EWOULDBLOCK)
+                {
+                    r.Complete(new IOException());
```

Comment:
Shouldn't the exception communicate the error that occurred?

Reply:
do we perhaps have any mechanism for translating error to the message? If we don't then I won't be able to provide much information here

Reply:
> do we perhaps have any mechanism for translating error to the message?

Yes, e.g. https://github.com/dotnet/corefx/blob/27fc4ecc80878f52feeac1d6bb23345ab65f97f5/src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs#L403.